File: modules/ticket_monitor.js
/**
 * Ticket Monitor Module
 * Handles monitoring for ticket availability and detecting adjacent seats
 */

const axios = require('axios');

class TicketMonitor {
  constructor() {
    this.lastCheckTime = null;
    this.lastAvailableTickets = null;
    // Check if we're in simulation mode
    this.simulationMode = process.env.BOT_SIMULATION === 'true';
  }

  /**
   * Check if tickets are available for the target match
   */
  async checkAvailability(targetUrl, proxy = null) {
    try {
      // If in simulation mode, randomly return true to simulate ticket detection
      if (this.simulationMode) {
        // 30% chance of detecting tickets
        return Math.random() < 0.3;
      }
      
      // Method 1: Direct page check using Playwright (more reliable)
      // This would be called from the main bot with a page instance
      
      // Method 2: HTTP request check (faster but less reliable)
      const config = {
        url: targetUrl,
        method: 'GET',
        timeout: 10000, // 10 second timeout
        headers: {
          'User-Agent': this.getRandomUserAgent(),
          'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
          'Accept-Language': 'en-US,en;q=0.5',
          'Accept-Encoding': 'gzip, deflate',
          'Connection': 'keep-alive',
          'Upgrade-Insecure-Requests': '1',
        }
      };
      
      // Add proxy if provided
      if (proxy) {
        config.proxy = this.parseProxy(proxy);
      }
      
      const response = await axios(config);
      
      // Analyze response for ticket availability indicators
      const responseBody = response.data;
      
      // Look for common indicators of available tickets
      const availabilityIndicators = [
        'tickets available',
        'buy now',
        'add to cart',
        'select seats',
        'ticket purchase',
        '"availability":"available"',
        '"in_stock":true',
        'data-available="true"'
      ];
      
      // Look for sold out indicators (negative check)
      const soldOutIndicators = [
        'sold out',
        'out of stock',
        'not available',
        '"availability":"sold_out"',
        '"in_stock":false',
        'data-available="false"'
      ];
      
      const lowerBody = responseBody.toLowerCase();
      
      // Check if sold out first
      const isSoldOut = soldOutIndicators.some(indicator => 
        lowerBody.includes(indicator)
      );
      
      if (isSoldOut) {
        return false;
      }
      
      // Check for availability indicators
      const hasAvailableTickets = availabilityIndicators.some(indicator => 
        lowerBody.includes(indicator)
      );
      
      // Also check for ticket quantity or price information
      const ticketPattern = /(\d+)\s*tickets?\s*available/i;
      const match = responseBody.match(ticketPattern);
      
      if (match) {
        const ticketCount = parseInt(match[1], 10);
        return ticketCount > 0;
      }
      
      return hasAvailableTickets;
    } catch (error) {
      // Network errors or timeouts don't necessarily mean no tickets
      // Log the error but don't treat it as "no tickets available"
      console.warn(`Monitoring error: ${error.message}`);
      return false; // Conservative approach - assume not available on error
    }
  }

  /**
   * Detect adjacent seats in the seating map
   * Returns a list of adjacent seat pairs
   */
  async detectAdjacentSeats(page) {
    if (this.simulationMode) {
      // In simulation mode, randomly generate adjacent seat pairs
      if (Math.random() < 0.4) { // 40% chance of finding adjacent seats
        return [
          {seat1: "A15", seat2: "A16", section: "North Stand"},
          {seat1: "B22", seat2: "B23", section: "East Stand"}
        ];
      } else {
        return [];
      }
    }
    
    // In real implementation, we would:
    // 1. Parse the seating map on the page
    // 2. Identify available seats
    // 3. Find adjacent pairs (same row, consecutive numbers)
    // 4. Return list of adjacent pairs
    
    return [];
  }

  /**
   * Get the current seating map from the page
   * Returns structured seating data
   */
  async getSeatingMap(page) {
    if (this.simulationMode) {
      // Simulate a seating map
      return {
        sections: [
          {
            name: "North Stand",
            rows: ["A", "B", "C", "D"],
            availableSeats: ["A15", "A16", "B22", "B23", "C10", "C12"]
          },
          {
            name: "East Stand",
            rows: ["E", "F", "G"],
            availableSeats: ["E5", "E6", "F12", "F13"]
          }
        ]
      };
    }
    
    // In real implementation, we would:
    // 1. Extract seating information from the page
    // 2. Parse the data into a structured format
    // 3. Return the seating map
    
    return {sections: []};
  }

  /**
   * Monitor tickets continuously and notify when available
   */
  async monitorContinuously(targetUrl, proxy = null, callback, intervalMs = 5000) {
    console.log(`Starting continuous monitoring for ${targetUrl}`);
    
    while (true) {
      try {
        const isAvailable = await this.checkAvailability(targetUrl, proxy);
        
        if (isAvailable) {
          console.log('TICKETS AVAILABLE! Notifying callback...');
          if (callback) {
            callback(true);
          }
          return true; // Stop monitoring after first detection
        } else {
          console.log('No tickets available yet. Waiting...');
        }
        
        // Wait before next check
        await this.sleep(intervalMs);
      } catch (error) {
        console.error(`Monitoring error: ${error.message}`);
        // Continue monitoring despite errors
        await this.sleep(intervalMs);
      }
    }
  }

  /**
   * Parse proxy string into axios proxy configuration
   */
  parseProxy(proxyString) {
    if (!proxyString) return null;
    
    try {
      const url = new URL(proxyString);
      return {
        protocol: url.protocol.replace(':', ''),
        host: url.hostname,
        port: url.port ? parseInt(url.port, 10) : 
               url.protocol === 'https:' ? 443 : 80,
        auth: url.username ? {
          username: url.username,
          password: url.password
        } : undefined
      };
    } catch (error) {
      console.warn(`Invalid proxy format: ${proxyString}`);
      return null;
    }
  }

  /**
   * Get a random realistic user agent
   */
  getRandomUserAgent() {
    const userAgents = [
      'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
      'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:109.0) Gecko/20100101 Firefox/121.0',
      'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
      'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.1 Safari/605.1.15',
      'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
    ];
    
    return userAgents[Math.floor(Math.random() * userAgents.length)];
  }

  /**
   * Utility function to sleep for specified milliseconds
   */
  sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
  }
}

module.exports = { TicketMonitor };

File: modules/ticket_monitor.py
"""
Ticket Monitor Module
Handles monitoring for ticket availability and detecting adjacent seats
"""

import asyncio
import os
import random
import time
import re
from utils.logger import Logger


class TicketMonitor:
    def __init__(self):
        self.logger = Logger()
        self.last_check_time = None
        self.last_available_tickets = None
        # Check if we're in simulation mode
        self.simulation_mode = os.environ.get('BOT_SIMULATION', 'false').lower() == 'true'

    async def check_availability(self, target_url, proxy=None):
        """Check if tickets are available for the target match"""
        self.logger.info(f"Checking ticket availability for {target_url}")
        
        # If in simulation mode, run the deterministic demo sequence
        if self.simulation_mode or os.getenv('SIMULATION_MODE', 'false').lower() == 'true':
            # 30% chance of detecting tickets in simulation mode
            is_available = random.random() < 0.3
            
            if is_available:
                # Return a simulated adjacent seat pair
                blocks = ["North Stand", "South Stand", "East Stand", "West Stand", "Block 32", "VIP Section"]
                rows = ["A", "B", "C", "D", "E", "10", "15", "20"]
                seat_num = random.randint(100, 200)
                
                return {
                    'available': True,
                    'block': random.choice(blocks),
                    'row': random.choice(rows),
                    'seat1': seat_num,
                    'seat2': seat_num + 1
                }
            else:
                return {'available': False}
        
        # In real implementation, we would use Playwright to check availability
        # This is a placeholder for the actual implementation
        try:
            # 1. Navigate to the target URL using Playwright
            # 2. Check for ticket availability
            # 3. Look for adjacent seat pairs
            # 4. Return True if adjacent seats are found
            pass
        except Exception as error:
            self.logger.error(f"Error checking ticket availability: {str(error)}")
            return {"success": False, "error": str(error)}

    async def find_adjacent_seats(self, page):
        """
        Find adjacent seats in the seating map
        Args:
            page: Playwright page object
        Returns:
            dict: Adjacent seat pair information or None if not found
        """
        if self.simulation_mode:
            # In simulation mode, randomly generate adjacent seat pairs
            if random.random() < 0.4:  # 40% chance of finding adjacent seats
                pairs = [
                    {"seat1": "A15", "seat2": "A16", "block": "North Stand", "ticket_ids": ["12345", "12346"]},
                    {"seat1": "B22", "seat2": "B23", "block": "East Stand", "ticket_ids": ["12347", "12348"]}
                ]
                selected_pair = random.choice(pairs)
                self.logger.info(f"Pair found: Block: {selected_pair['block']}, Row: {selected_pair['seat1'][0]}, Seats: {selected_pair['seat1']} & {selected_pair['seat2']} (ticket_ids: {','.join(selected_pair['ticket_ids'])})")
                return [selected_pair]
            else:
                # Log top 5 nearest options when no pairs found
                single_seats = [
                    {"block": "North Stand", "row": "A", "seat": "A15", "ticket_id": "12345"},
                    {"block": "North Stand", "row": "A", "seat": "A17", "ticket_id": "12349"},
                    {"block": "East Stand", "row": "B", "seat": "B22", "ticket_id": "12347"},
                    {"block": "East Stand", "row": "B", "seat": "B24", "ticket_id": "12350"},
                    {"block": "West Stand", "row": "C", "seat": "C10", "ticket_id": "12351"}
                ]
                self.logger.warn("No adjacent pair available. Top single seats:")
                for seat in single_seats[:5]:
                    self.logger.warn(f"  Block {seat['block']} Row {seat['row']} Seat {seat['seat']} (ticket_id: {seat['ticket_id']})")
                return []
        
        # In real implementation, we would:
        # 1. Parse the seating map on the page
        # 2. Identify available seats
        # 3. Find adjacent pairs (same row, consecutive numbers)
        # 4. Return list of adjacent pairs
        
        return []

    def parse_seat_map(self, seat_data):
        """
        Parse seat map data to find adjacent seats
        Args:
            seat_data: Raw seat map data from API or page
        Returns:
            list: List of adjacent seat pairs
        """
        if not seat_data:
            return []
            
        # Group seats by row
        rows = {}
        for seat in seat_data:
            # Extract row (first letter) and seat number
            match = re.match(r'^([A-Z])(\d+)$', seat)
            if match:
                row, number = match.groups()
                if row not in rows:
                    rows[row] = []
                rows[row].append((int(number), seat))
        
        # Find adjacent pairs in each row
        pairs = []
        for row, seat_numbers in rows.items():
            # Sort by seat number
            seat_numbers.sort()
            
            # Find consecutive numbers
            for i in range(len(seat_numbers) - 1):
                num1, seat1 = seat_numbers[i]
                num2, seat2 = seat_numbers[i + 1]
                
                if num2 == num1 + 1:  # Adjacent seats
                    pairs.append({
                        "seat1": seat1,
                        "seat2": seat2,
                        "row": row,
                        "numbers": [num1, num2]
                    })
        
        return pairs
